docs/development/design/embed-api.rst

```diff
@@ -0,0 +1,211 @@
+Embed APIv3
+===========
+
+The Embed API allows to embed content from documentation pages in other sites.
+It has been treated as an *experimental* feature without public documentation or real applications,
+but recently it started to be used widely (mainly because we created a Sphinx extension).
+
+The main goal of this document is to design a new version of the Embed API to be more user friendly,
+make it more stable over time, support documentation pages not hosted at Read the Docs,
+and remove some quirkiness that makes it hard to maintain and difficult to use.
+
+.. note::
+
+   This work is part of the `CZI grant`_ that Read the Docs received.
+
+.. _CZI grant: https://blog.readthedocs.com/czi-grant-announcement/
+
+.. contents::
+   :local:
+   :depth: 2
+
+
+Current implementation
+----------------------
+
+The current implementation of the API is partially documented in :doc:`/guides/embedding-content`.
+It has some known problems:
+
+* There are different ways of querying the API: ``?url=`` (generic) and ``?doc=`` (relies on Sphinx's specific concept)
+* Doesn't support MkDocs
+* Lookups are slow (~500 ms)
+* IDs returned aren't well formed (like empty IDs ``"headers": [{"title": "#"}]``)
+* The content is always an array of one element
+* It tries different variations of the original ID
+* It doesn't return valid HTML for definition lists (``dd`` tags without a ``dt`` tag)
+
+
+Goals
+-----
+
+Considering the problems mentioned in the previous section,
+the inclusion of new features and the definition of a contract that works the same for all,
+this document set the following goals for the new version of this endpoint:
+
+* Support external documents hosted outside Read the Docs
+* Do not depend on Sphinx ``.fjson`` files
+* Query and parse the ``.html`` file directly (from our storage or from an external request)
+* Rewrite all links returned in the content to make them absolute
+* Always return valid HTML structure
+* Delete HTML tags from the original document if needed
+* Support ``?nwords=`` and ``?nparagraphs=`` to return chunked content
+* Require a valid HTML ``id`` selector
+* Handle special cases for particular doctools (e.g. Sphinx requires to return the ``.parent()`` element for ``dl``)
+* Make explicit the client is asking to handle the special cases (e.g. send ``?doctool=sphinx&version=4.0.1``)
```

Comment:
I think we voted against handling special cases, and as mentioned in my other comment this isn't a special case for sphinx, but following the semantic representation of html tags.

Reply:
This is definitely a special case considering how our contract was defined: "return the HTML tag with the `id=` requested". However, this case is "return the **parent** of HTML tag with the `id=` requested". It may not be a Sphinx-specific special case, but we need a way to communicate the backend we want to benefit from that extra manipulation: "return the parent _and_ remove the siblings"

I see this as some "magic" the backend is doing for us and, as a client, we will need a way to opt-in/out to this "magic". Adding a config to opt-in/out will make this endpoint maintainable overtime when we discover more similar cases.

Reply:
We are committed to return valid html, so I don't see that as magic, is just following a well-defined case for a html tag.

Reply:
No, the contract is not about valid/invalid HTML. We are committing ourselves to return the HTML tag for the `id` requested. Any other response that is outside this contract should be managed as a special case.

Also, we can't commit ourselves to "always return valid HTML" because we don't know if the HTML we are parsing is valid or not. In fact, I just removed that from the "Goals" section.

I updated the design doc with a new section, "The API contract", that exposes our commitment.

Reply:
What I mean is that we are committed to return valid html for the well-defined cases, we don't try to fix invalid html, we just don't convert valid html into invalid html.

Reply:
I think you are both saying the same thing. We are returning the HTML as it exists on the page. It isn't valid or invalid, it's modified or unmodified, and we plan to serve it unmodified.

Reply:
> I think you are both saying the same thing. We are returning the HTML as it exists on the page.

Yes... and no 😄  We are saying the same about valid/invalid HTML responses.

Apart from that, I'm also saying that I want to have a way to specify "special cases" that don't follow _exctly_ the contract we defined.